Add YOLO11 TensorRT CUDA inference example

shikhindahikar opened the pull request:
<!--
Thank you 🙏 for your contribution to [Ultralytics](https://www.ultralytics.com/) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. Check for Existing Contributions: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. Link Related Issues: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. Elaborate Your Changes: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. Ultralytics Contributor License Agreement (CLA): To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    I have read the CLA Document and I sign the CLA

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Your adherence to these guidelines ensures a faster and more effective review process.
-->


## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Adds a complete C++ example for running YOLO11 inference with TensorRT and CUDA, including build files, usage guide, and a working demo app. ⚡️🖥️

### 📊 Key Changes
- New example: `examples/YOLO11-TensorRT-CUDA-Inference/`
  - `main.cu`: End-to-end TensorRT inference pipeline (engine deserialization, CUDA buffers, preprocessing, inference, simple drawing of detections).
  - `TRTHandle.h`: Helper utilities for TensorRT logging, CUDA error handling, and dynamic I/O buffer management.
  - `CMakeLists.txt`: Cross-platform build config (Windows/Linux) with OpenCV and TensorRT linking.
  - `README.md`: Step-by-step instructions to export YOLO11 to ONNX, convert to TensorRT engine with `trtexec`, build with CMake, and run the demo.
- Demonstrates preprocessing (letterbox + `blobFromImage`) and basic output parsing to draw boxes on frames.

### 🎯 Purpose & Impact
- Faster deployment: Provides a ready-to-run TensorRT CUDA example for high-performance YOLO11 inference on NVIDIA GPUs. 🚀
- Developer-friendly: Clarifies the export path (YOLO11 → ONNX → TensorRT engine), making integration into C++/CUDA pipelines easier.
- Cross-platform setup: CMake config includes both Windows and Linux linking patterns for TensorRT and OpenCV.
- Practical starting point: Helps teams bootstrap custom C++ TensorRT apps for production; users can extend post-processing (e.g., NMS) and adapt to their engine I/O shapes.
- Notes/caveats:
  - Users must update placeholder paths for OpenCV and TensorRT in CMake.
  - Example assumes a simplified output layout (6 values per detection); adjust parsing to match your exported YOLO11 engine’s actual outputs.
  - Requires TensorRT 10.12+, CUDA, and OpenCV installed.

shikhindahikar committed:
init

shikhindahikar committed:
init

shikhindahikar committed:
Added:
- the inference feature
- build system with CMake
- changed the directory name to a more appropriate one

shikhindahikar committed:
Added:
- the inference feature
- build system with CMake
- changed the directory name to a more appropriate one

shikhindahikar committed:
Added: Updated README, letterbox function to main.cu, Fixed the frame transfer from the GPU to CPU in main.cu, Updated the CMakeLists.txt for Linux as well

shikhindahikar committed:
Added: Updated README, letterbox function to main.cu, Fixed the frame transfer from the GPU to CPU in main.cu, Updated the CMakeLists.txt for Linux as well

shikhindahikar commented:
I have read the CLA Document and I sign the CLA

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant commented:
👋 Hello @shikhindahikar, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help speed things up—an Ultralytics engineer will review and assist you shortly. In the meantime, please review the checklist below to ensure a smooth merge:

- ✅ Define a Purpose: Clearly explain the purpose of your fix or feature in your PR description, and link to any relevant issues(https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ Synchronize with Source: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ Ensure CI Checks Pass: Verify all Ultralytics Continuous Integration (CI)(https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ Update Documentation: Update the relevant documentation(https://docs.ultralytics.com/) for any new or modified features.
- ✅ Add Tests: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ Sign the CLA: Please ensure you have signed our Contributor License Agreement(https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ Minimize Changes: Limit your changes to the minimum necessary for your bug fix or feature addition. "It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is." — Bruce Lee

For more guidance, please refer to our Contributing Guide(https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

PR-specific notes to help expedite review:
- Branding consistency: Please update any remaining references from “YOLOv11” to “YOLO11” across filenames, README content, and comments to match Ultralytics naming conventions ✅
- Example structure: The new example path `examples/YOLO11-TensorRT-CUDA-Inference/` looks good. Ensure large binary assets (e.g., `.engine` files) are not committed and are gitignored where applicable.
- CMake polish: Consider replacing hardcoded placeholder paths with cache variables/options (e.g., `-DTENSORRT_ROOT=...`, `-DOpenCV_DIR=...`) to improve portability. Also note `set(CMAKE_CUDA_STANDARD 14)` appears twice—please deduplicate.
- Headers/logging: In `TRTHandle.h`, `sample::gLogError` is referenced but not defined in the snippet. Please either provide the definition or replace those macros with standard logging/streams to avoid unresolved references.
- Tensor shapes/I/O: Thanks for calling out output parsing assumptions. A brief note in the README guiding users to verify their engine’s I/O names, shapes, and dtype at runtime (as you print) would help prevent confusion.
- Error handling: Great use of CUDA stream syncs. You might also guard against empty `cv::imread` results and validate `argv` parsing for width/height if exposed later.
- Documentation: Since this is a full example, please ensure the included README is aligned with Ultralytics docs terminology and installation guidance, and that it references the YOLO11 export pathway consistently.
- Tests/CI: If feasible, add a lightweight build check or static check for the example to help CI signal breakages early (no need to run inference in CI).

Once the above are addressed and CI is green, this looks like a helpful addition for users deploying YOLO11 with TensorRT and CUDA ⚡️🖥️

shikhindahikar committed:
Update README.md

Changed all YOLOv11 references to YOLO11 for branding compliance

Signed-off-by: Shikhin Dahikar <31371540+shikhindahikar@users.noreply.github.com>

shikhindahikar committed:
Update CMakeLists.txt

Removing redundant variable set in CMake file

Signed-off-by: Shikhin Dahikar <31371540+shikhindahikar@users.noreply.github.com>

shikhindahikar committed:
Removing the unused macros

shikhindahikar commented:
recheck

glenn-jocher commented:
Thanks for the update—your boxes look correct on the 640×640 letterboxed image because the postprocess hasn’t undone the padding/scale. To map detections back to the original image, subtract the letterbox padding and divide by the scale ratio (use the exact `r, dw, dh` you used in preprocessing):

```cpp
// Undo letterbox for xyxy boxes produced at net_w x net_h (e.g., 640x640)
inline void undo_letterbox_xyxy(float& x1, float& y1, float& x2, float& y2,
                                int img_w, int img_h, int net_w, int net_h) {
    const float r = std::min(net_w / static_cast<float>(img_w),
                             net_h / static_cast<float>(img_h));
    const float new_w = img_w * r, new_h = img_h * r;
    const float dw = (net_w - new_w) * 0.5f;
    const float dh = (net_h - new_h) * 0.5f;

    x1 = (x1 - dw) / r;  y1 = (y1 - dh) / r;
    x2 = (x2 - dw) / r;  y2 = (y2 - dh) / r;

    x1 = std::max(0.f, std::min(x1, static_cast<float>(img_w)));
    y1 = std::max(0.f, std::min(y1, static_cast<float>(img_h)));
    x2 = std::max(0.f, std::min(x2, static_cast<float>(img_w)));
    y2 = std::max(0.f, std::min(y2, static_cast<float>(img_h)));
}
```

If your engine outputs `cx, cy, w, h`, convert to `xyxy` before calling the function:
```cpp
x1 = cx - 0.5f * w;  y1 = cy - 0.5f * h;
x2 = cx + 0.5f * w;  y2 = cy + 0.5f * h;
```

Optional: if you want the engine to return final `Nx6 [x1,y1,x2,y2,score,cls]`, export with NMS fused using `model.export(format="engine", nms=True)`; see the TensorRT export arguments in the [TensorRT integration docs](https://docs.ultralytics.com/integrations/tensorrt/).

shikhindahikar committed:
Added:
- post processing function to normalize back to the original scale
- Now shows the original image with the bounding boxes

shikhindahikar committed:
Updated README to include alternative methods of exporting model to engine

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

shikhindahikar committed:
Updated README to include a note for INT8 model calibration

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

shikhindahikar commented on examples/YOLO11-TensorRT-CUDA-Inference/main.cu:
```diff
@@ -0,0 +1,221 @@
+        }
+        std::cout << ", Data type: " << static_cast<int32_t>(dtype)
+                << std::endl;
+    }
```
Hi @onuralpszr could you or anyone else can please help me with finishing this PR?

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

shikhindahikar commented:
I had lot of merge conflicts in this PR because of auto formatting bot perhaps I think, I have tried to resolve it but also ended up modifying build.py by mistake, so made another commit to resolve that. I hope this is okay?

shikhindahikar commented:
Rebased to main and force pushed. This should work.

onuralpszr commented:
> Rebased to main and force pushed. This should work.

Thank you, I will test it.

shikhindahikar commented:
hi @onuralpszr @glenn-jocher  this PR is pending for quite a while now. Could you please check?

shikhindahikar committed:
Update CMakeLists.txt

Removing redundant variable set in CMake file

Signed-off-by: Shikhin Dahikar <31371540+shikhindahikar@users.noreply.github.com>

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher commented:
Thank you @shikhindahikar for the persistence, hardware work, and detailed TensorRT/CMake example. The example itself is useful and the effort here is absolutely appreciated. We’re closing this branch because it has accumulated a large set of unrelated package changes and no longer provides a safe reviewable integration unit against current `main`; this is a branch-shape decision, not a rejection of the example. A fresh example-only PR with the known-good environment, exact export/build commands, and no package-source changes would be much easier to validate and would be welcome. Thanks again for investing the time in this. 🙏
